engine/crates/runtime-local/src/rate_limiting/in_memory/key_based.rs

```diff
@@ -34,48 +37,72 @@ impl RateLimiterContext for RateLimitingContext {
     }
 }
 
-#[derive(Default)]
 pub struct InMemoryRateLimiter {
-    inner: HashMap<String, governor::DefaultKeyedRateLimiter<usize>>,
+    limiters: Arc<RwLock<HashMap<String, governor::DefaultKeyedRateLimiter<usize>>>>,
 }
 
 impl InMemoryRateLimiter {
-    pub fn runtime(config: KeyedRateLimitConfig<'_>) -> RateLimiter {
-        let mut limiter = Self::default();
+    pub fn runtime(
+        config: KeyedRateLimitConfig,
+        mut updates: mpsc::Receiver<HashMap<String, GraphRateLimit>>,
+    ) -> RateLimiter {
+        let mut limiters = HashMap::new();
 
         // add subgraph rate limiting configuration
-        for (name, rate_limit_config) in config.rate_limiting_configs {
-            limiter = limiter.with_rate_limiter(name, rate_limit_config);
+        for (name, config) in config.rate_limiting_configs {
+            let Some(limiter) = create_limiter(config) else {
+                continue;
+            };
+
+            limiters.insert(name.to_string(), limiter);
         }
 
-        RateLimiter::new(limiter)
-    }
+        let limiters = Arc::new(RwLock::new(limiters));
+        let limiters_copy = limiters.clone();
+
+        tokio::spawn(async move {
+            while let Some(updates) = updates.recv().await {
+                let mut limiters = limiters_copy.write().unwrap();
+
+                for (name, config) in updates {
+                    let Some(limiter) = create_limiter(config) else {
+                        continue;
+                    };
+
+                    limiters.insert(name.to_string(), limiter);
+                }
+            }
+        });
 
-    pub fn with_rate_limiter(mut self, key: &str, rate_limit_config: GraphRateLimit) -> Self {
-        let quota = (rate_limit_config.limit as u64)
-            .checked_div(rate_limit_config.duration.as_secs())
-            .expect("rate limiter with invalid per second quota");
-
-        self.inner.insert(
-            key.to_string(),
-            governor::RateLimiter::keyed(Quota::per_second(
-                NonZeroU32::new(quota as u32).expect("rate limit duration cannot be 0"),
-            )),
-        );
-        self
+        RateLimiter::new(Self { limiters })
     }
 }
 
+fn create_limiter(rate_limit_config: GraphRateLimit) -> Option<governor::DefaultKeyedRateLimiter<usize>> {
```

Comment:
:bug: **Bug**
The `create_limiter` function does not handle the case where the `rate_limit_config.duration` is zero, which could lead to a division by zero error when calculating the quota. This should be checked before performing the division.

```suggestion
let Some(quota) = (rate_limit_config.limit as u64).checked_div(rate_limit_config.duration.as_secs()).filter(|&d| d > 0) else { tracing::error!(target: GRAFBASE_TARGET, "the duration for rate limit cannot be zero"); return None; };
```